### PR TITLE
Add PCLVisualizer::containsCloud.

### PR DESCRIPTION
**Pull request by sotte**
Hey,

I "needed" the following function and it's just two lines of code. It basically allows you to check the PCLVisualizer contains the cloud with the given id independent of the type of the cloud.

Let me know what you think.

-----

`containsCloud` allows to check if the viewer already contains a cloud with the
given id.

`updatePointCloud` also checks this AND updates the point cloud, however,
in many scenarios only the check is important.
`containsCloud` is much clearer and only requires the id of the cloud.

Why `containsCloud`:
In many tutorials the user keeps track of the added pointclouds by hand, i.e.,
flags like `graph_added` are used.
This is necessary because the `updatePointCloud` cannot be used.
`containsCloud` solves this issue.

### CODE REVIEW

File: visualization/include/pcl/visualization/pcl_visualizer.h [hunk 1]
```diff
@@ -579,6 +579,17 @@ namespace pcl
                    double r = 1.0, double g = 1.0, double b = 1.0,
                    const std::string &id = "", int viewport = 0);
 
+        /** \brief Check if the clound with the given id was already added to this vizualizer.
```
Comment (VictorLamoine) [inline visualization/include/pcl/visualization/pcl_visualizer.h]: cloud not clound :wink:

Comment (sotte) [inline visualization/include/pcl/visualization/pcl_visualizer.h]: And this is why reviews are a good idea, thanks! :)

**Comment by taketwo** [comment]
:+1:

There is one subtlety that's worth documenting. For some reason when you add a polygon mesh (with [`addPolygonMesh()`](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/include/pcl/visualization/impl/pcl_visualizer.hpp#L1565)), the created actor ends up in `cloud_actor_map_`. Therefore your check will be give false positive if a polygon mesh with the given id already exists.

I would add a note/warning to the docstring to avoid confusion.

**Comment by sotte** [comment]
I searched a bit  more and pretty much everything you add to the viewer ends up in the `cloud_actor_map_`, i.e., all functions that start with `add` extend the `cloud_actor_map_`.
Here is an imcomplete list:
- `addPointCloud`
- `addPolyModelFromPolyData`
- `addTextureMesh`

I guess the underlying problem is that the concept of the `cloud_actor_map_` is not really explained anywhere and the name implies that it only deals with clouds, which is not the case.

Anyway, should we mention previously named functions as well? Or should we rename the function? `containsId`, `containsInput`, `containsXXX`?

**Comment by taketwo** [comment]
Maybe this function should also check `shape_actor_map_`?

**Comment by sotte** [comment]
Oh man, I didn't even know that `shape_actor_map_` exists. There is also `coordinate_actor_map_`.

I don't know what is the best way to proceed. How about this:

- `containsCloud(const std::string& id)` only checks the `cloud_actor_map_`
- `containsShape(const std::string& id)` only checks the `shape_actor_map_`
- `containsCoordinate(const std::string& id)` only checks the `coordinate_actor_map_` (Is this function really necessary? I never worked with coordinates.)
- `contains(const std::string& id)` checks all of the above, basically checking if something with the given id was added.

**Comment by taketwo** [comment]
Or we can enforce uniqueness of the name independently of the particular actor map where it ends up. Then only a single `contains` function will be needed. What do you think?

**Comment by sotte** [comment]
This would mean that every time we want to add "stuff" we have to check if the id aleady exists in one of the three actor maps, right?

This would require quite a bit of refactoring (nothing hard but quite a few lines of code). My two-liner is turning into something bigger but I guess I could implement it.

Would this somehow influence the backward compatibility?

**Comment by taketwo** [comment]
The refactoring will amount to changing lines like:

```cpp
ShapeActorMap::iterator am_it = shape_actor_map_->find (id);
if (am_it != shape_actor_map_->end ())
{
   // ...
```

with which every `add` function starts already to just a call to your new `contains` function.

> Would this somehow influence the backward compatibility?

Well may give problems to those who used objects of different types with the same name. (But I have hard times imagining why someone would add a point cloud called "cloud" and then a sphere also called "cloud"). Anyway, a warning message will be printed, so it will be relatively easy to solve the problem for those who are affected.

**Comment by sotte** [comment]
OK, I'll implement it...

**Comment by sotte** [comment]
I just found https://github.com/PointCloudLibrary/pcl/blob/master/visualization/src/pcl_visualizer.cpp#L780.
There seems to be a certain semantic for ids with the same name in ShapeActorMap and CloudActorMap. Now I'm wondering if we don't break anything by requiering uniqueness.

**Comment by taketwo** [comment]
I don't think so. Check [this commit message](https://github.com/PointCloudLibrary/pcl/commit/cdf286e6c2d892e8c265260c49f7785bbbf0acbf). It seems that there was/is quite a jumble with what goes where (to which actor map). For example, polygon meshes end up in cloud map, whereas polygons go to the shape map...

**Comment by sotte** [comment]
Some addPointCloud function do multiple things. Here is an example:

```cpp
pcl::visualization::PCLVisualizer::addPointCloud (
   const typename pcl::PointCloud<PointT>::ConstPtr &cloud,
   const GeometryHandlerConstPtr &geometry_handler,
   const std::string &id, int viewport)
{
  CloudActorMap::iterator am_it = cloud_actor_map_->find (id);
  if (am_it != cloud_actor_map_->end ())
  {
    am_it->second.geometry_handlers.push_back (geometry_handler);
    return (true);
  }
  PointCloudColorHandlerCustom<PointT> color_handler (cloud, 255, 255, 255);
  return (fromHandlersToScreen (geometry_handler, color_handler, id, viewport, cloud->sensor_origin_, cloud->sensor_orientation_));
}
```

If id was not added yet, the data is added.
If id was found, the handler is updated.

Should the later case not be handled in the update function?

**Comment by taketwo** [comment]
Hm... maybe. But that's a different issue, isn't it?

Could you please squash the first two commits? Also please correct a few whitespace/line break issues in the last commit (lines 85, 124, 614).

:+1: for the elaborate commit message!

**Comment by sotte** [comment]
I just wanted to mention the issue though.

I squashed the two commits. But I'm not sure about the whitespaces. Looks ok to me. What exactly do you mean?

File: visualization/include/pcl/visualization/impl/pcl_visualizer.hpp [hunk 2]
```diff
@@ -82,12 +82,9 @@ pcl::visualization::PCLVisualizer::addPointCloud (
   const PointCloudGeometryHandler<PointT> &geometry_handler,
   const std::string &id, int viewport)
 {
-  // Check to see if this ID entry already exists (has it been already added to the visualizer?)
-  CloudActorMap::iterator am_it = cloud_actor_map_->find (id);
-
-  if (am_it != cloud_actor_map_->end ())
+  if ( contains(id))
```
Comment (taketwo) [inline visualization/include/pcl/visualization/impl/pcl_visualizer.hpp]: Extra space before and missing space after `contains`

File: visualization/include/pcl/visualization/impl/pcl_visualizer.hpp [hunk 3]
```diff
@@ -126,12 +121,9 @@ pcl::visualization::PCLVisualizer::addPointCloud (
   const PointCloudColorHandler<PointT> &color_handler,
   const std::string &id, int viewport)
 {
-  // Check to see if this ID entry already exists (has it been already added to the visualizer?)
-  CloudActorMap::iterator am_it = cloud_actor_map_->find (id);
-
-  if (am_it != cloud_actor_map_->end ())
+  if (contains(id))
```
Comment (taketwo) [inline visualization/include/pcl/visualization/impl/pcl_visualizer.hpp]: Missing space after `contains`

File: visualization/include/pcl/visualization/impl/pcl_visualizer.hpp [hunk 4]
```diff
@@ -632,13 +611,13 @@ pcl::visualization::PCLVisualizer::addSphere (const PointT &center, double radiu
 template<typename PointT> bool
 pcl::visualization::PCLVisualizer::updateSphere (const PointT &center, double radius, double r, double g, double b, const std::string &id)
 {
-  // Check to see if this ID entry already exists (has it been already added to the visualizer?)
-  ShapeActorMap::iterator am_it = shape_actor_map_->find (id);
-  if (am_it == shape_actor_map_->end ())
+  if (contains (id)) {
```
Comment (taketwo) [inline visualization/include/pcl/visualization/impl/pcl_visualizer.hpp]: `{` should go on the next line

**Comment by taketwo** [comment]
Thanks! I left inline comments.

**Comment by sotte** [comment]
Thanks! Updated.

**Pull request merged**